kinder/pkg/cluster/manager/actions/kubeadm-config.go

```diff
@@ -272,14 +325,101 @@ func getKubeadmConfig(c *status.Cluster, n *status.Node, data kubeadm.ConfigData
 	// if the node is the bootstrap control plane, then all the objects used as init time
 	if n == c.BootstrapControlPlane() {
 		return selectYamlFramentByKind(patched,
+			"MasterConfiguration", //v1alpha2
 			"ClusterConfiguration",
 			"InitConfiguration",
 			"KubeletConfiguration",
 			"KubeProxyConfiguration"), nil
 	}
 
 	// otherwise select only the JoinConfiguration
-	return selectYamlFramentByKind(patched, "JoinConfiguration"), nil
+	return selectYamlFramentByKind(patched,
+		"NodeConfiguration", //v1alpha2
+		"JoinConfiguration",
+	), nil
+}
+
+func createDiscoveryFile(c *status.Cluster, n *status.Node, discoveryMode DiscoveryMode) error {
+	// the discovery file is a kubeaconfig file, so for sake of semplicity in setting up this test,
+	// we are using the admin.conf file created by kubeadm on the bootstrap control plane node
+	// as a starting point (e.g. it already contains the necessary server address/server certificate)
+	// IMPORTANT. Don't do this in production, admin.conf contains cluster-admin credentials.
+	lines, err := c.BootstrapControlPlane().Command(
+		"cat", "/etc/kubernetes/admin.conf",
+	).Silent().RunAndCapture()
+	if err != nil {
```

Comment:
can be:
`if err != nil || len(lines) == 0 {`

Reply:
if `err==nil` and `len(lines) == 0`, which happen e.g. if you try to create discovery file before init, then you call `errors.Wrapf(err, ...` with `err==nil` which sadly returns `nil` instead of an error. 
So we have to break conditions in two